go/vt/vtadmin/cache/cache.go

```diff
@@ -0,0 +1,262 @@
+//			// Fetch schema based on fields in `req`.
+//			// If err is nil, the backfilled schema will be added to the cache.
+//			return cluster.fetchSchema(ctx, req)
+//		})
+//
+type Cache[Key Keyer, Value any] struct {
+	cache     *cache.Cache
+	fillcache *cache.Cache
+
+	ctx       context.Context
+	cancel    context.CancelFunc
+	wg        sync.WaitGroup
+	backfills chan *backfillRequest[Key]
+
+	fillFunc func(ctx context.Context, k Key) (Value, error)
+
+	cfg Config
+}
+
+// New creates a new cache with the given backfill func. When a request is
+// enqueued (via EnqueueBackfill), fillFunc will be called with that request.
+func New[Key Keyer, Value any](fillFunc func(ctx context.Context, req Key) (Value, error), cfg Config) *Cache[Key, Value] {
+	if cfg.BackfillEnqueueWaitTime <= 0 {
+		log.Warningf("BackfillEnqueueWaitTime (%v) must be positive, defaulting to %v", cfg.BackfillEnqueueWaitTime, DefaultBackfillEnqueueWaitTime)
+		cfg.BackfillEnqueueWaitTime = DefaultBackfillEnqueueWaitTime
+	}
+
+	if cfg.BackfillRequestTTL <= 0 {
+		log.Warningf("BackfillRequestTTL (%v) must be positive, defaulting to %v", cfg.BackfillRequestTTL, DefaultBackfillRequestTTL)
+		cfg.BackfillRequestTTL = DefaultBackfillRequestTTL
+	}
+
+	c := &Cache[Key, Value]{
+		cache:     cache.New(cfg.DefaultExpiration, cfg.CleanupInterval),
+		fillcache: cache.New(cfg.BackfillRequestDuplicateInterval, time.Minute),
+		backfills: make(chan *backfillRequest[Key], cfg.BackfillQueueSize),
+		fillFunc:  fillFunc,
+		cfg:       cfg,
+	}
+
+	c.ctx, c.cancel = context.WithCancel(context.Background())
+
+	c.wg.Add(1)
+	go c.backfill() // TODO: consider allowing N backfill threads to run, configurable
+
+	return c
+}
+
+// Add adds a (key, value) to the cache directly, following the semantics of
+// (github.com/patrickmn/go-cache).Cache.Add.
+func (c *Cache[Key, Value]) Add(key Key, val Value, d time.Duration) error {
+	return c.add(key.Key(), val, d)
+}
+
+func (c *Cache[Key, Value]) add(key string, val Value, d time.Duration) error {
+	// Record the time we last cached this key, to check against
+	c.fillcache.Set(key, struct{}{}, cache.DefaultExpiration)
+	// Then cache the actual value.
+	return c.cache.Add(key, val, d)
+}
+
+// Get returns the Value stored for the key, if present in the cache. If the key
+// is not cached, the zero value for the given type is returned, along with a
+// boolean to indicated presence/absence.
+func (c *Cache[Key, Value]) Get(key Key) (Value, bool) {
+	v, exp, ok := c.cache.GetWithExpiration(key.Key())
+	if !ok || (!exp.IsZero() && exp.Before(time.Now())) {
+		var zero Value
+		return zero, false
+	}
+
+	return v.(Value), ok
+}
+
+// EnqueueBackfill submits a request to the backfill queue.
+func (c *Cache[Key, Value]) EnqueueBackfill(k Key) bool {
+	req := &backfillRequest[Key]{
+		k:           k,
+		requestedAt: time.Now().UTC(),
+	}
+
+	ctx, cancel := context.WithTimeout(c.ctx, c.cfg.BackfillEnqueueWaitTime)
+	defer cancel()
```

Comment:
Since you're not passing the context anywhere, this is better modeled as a `time.After`.